### ISSUE TITLE
Send string (Variant vs PoolStringArray) over the stream.

### ISSUE THREAD

**Issue by Fulkman** [godotengine/godot#17230]
**Godot version:**
Godot 3.0.stable.official

**OS/device including version:**
OS X 10.13.x

**Issue description:**
String into the stream:
`streamPeerTCP.put_var('test')` sends
`<Buffer ... 04 00 00 00 74 65 73 74 />` - **OK**

PoolStringArray into the stream:
`streamPeerTCP.put_var(PoolStringArray(['test']))` sends
`<Buffer ... 05 00 00 00 74 65 73 74 00 xx xx xx />` - **???**

x - irrelevant bytes

There is inconsistency. Why send data as `PoolStringArray` are suffixed with `\00` and string length is +1? Is there any reason for that behavior?

**Comment by AlexHolly** [comment]
I think it has something to do with this http://docs.godotengine.org/en/3.0/tutorials/misc/binary_serialization_api.html

**Comment by Fulkman** [comment]
I'm full aware of this page. My question was Why not How.

I can send buffer from my server: (without extra `\00` byte)
`<Buffer 10000000 17000000 01000000 04000000 74657374 />`

```
var variant = packetPeerStream.get_var();
printt(typeof(variant), variant) // 23	[test]
``` 

Work as expected.

**Comment by Faless** [comment]
> There is inconsistency. Why send data as PoolStringArray are suffixed with \00 and string length is +1? Is there any reason for that behavior?

As mentioned in the link, all Godot serialized object are aligned to the 4th byte (sizes in bits are multiple of 32). When the packet is shorter, empty (`0x00`) padding is added.

This is done for performance reasons when copying the buffer (and possibly compatibility with exotic non-BSD sockets). This is very common practice in network protocols.

> I can send buffer from my server: (without extra \00 byte)

That was a bug, and a potential security issue, fixed in #20558 .

Closing as it now works as intended.
